### 2. В команде есть два стажера-аналитика. Правильный ответ каждый из стажеров получает в 14 из 17 случаев. Чтобы быть более уверенным в важном решении, менеджер решил дать одну и ту же задачу сразу обоим аналитикам: если оба получают одинаковый ответ, то менеджер его использует, а если ответы разные - выберет один из них наугад. Насколько такой способ повышает шансы менеджера принять верное решение?

In [1]:
import numpy as np

In [2]:
n = 10000000 # количество решений
x = np.random.choice([0,1], n ,p=[3/17, 14/17]) # n - решений первого стажера-аналитика
y = np.random.choice([0,1], n ,p=[3/17, 14/17]) # n - решений второго стажера-аналитика

In [3]:
ver_true_x = x[x==1].shape[0] /n
ver_false_x = x[x==0].shape[0] / n

ver_true_y = y[y==1].shape[0] /n
ver_false_y = y[y==0].shape[0] / n

In [4]:
print('Вероятность правильного решения \n1 стажер - аналитик \n\tправильно: {:.2%} неправильно: {:.2%}\n2 стажер - аналитик \n\tправильно: {:.2%} неправильно: {:.2%}'.format(
    ver_true_x, ver_false_x, 
    ver_true_y, ver_false_y))

Вероятность правильного решения 
1 стажер - аналитик 
	правильно: 82.36% неправильно: 17.64%
2 стажер - аналитик 
	правильно: 82.36% неправильно: 17.64%


In [5]:
print('Вероятность правильного решения:')
print('1 стажер - аналитик \n\tправильно: {:.2%} неправильно: {:.2%}'.format(ver_true_x, ver_false_x))
print('2 стажер - аналитик \n\tправильно: {:.2%} неправильно: {:.2%}'.format(ver_true_y, ver_false_y))

Вероятность правильного решения:
1 стажер - аналитик 
	правильно: 82.36% неправильно: 17.64%
2 стажер - аналитик 
	правильно: 82.36% неправильно: 17.64%


##### Рассмотрим  первый вариант при котором стажеры-аналитики совершают одинаковые ошибки при решении задач, тогда в случае если оба решения совпадает получается ошибка (примерно 4.4%), когда они оба решили неверно, но решения совпали

In [6]:
# вероятность что решения совпадают: оба решили правильно + оба допустили одинаковую ошибку (два одинаковых решения)
сhance_x_eq_y = ver_true_x*ver_true_y + ver_false_x*ver_false_y
# вероятность что решения не совпадают: первый решил верно второий нет (два разных решения)
сhance_x_noeq_y = ver_true_x*ver_false_y + ver_false_x*ver_true_y

In [7]:
print('вероятность того, что решения совпадают: {:.2%}\nрешения не совпадают: {:.2%}'.format(сhance_x_eq_y, сhance_x_noeq_y))

вероятность того, что решения совпадают: 70.94%
решения не совпадают: 29.06%


In [8]:
all_x_eq_y = x[x==y]
all_x_noeq_y = x[x!=y]

In [9]:
x_eq_y_right = all_x_eq_y[all_x_eq_y == 1].shape[0] / all_x_eq_y.shape[0]
x_eq_y_mistake = all_x_eq_y[all_x_eq_y == 0].shape[0] / all_x_eq_y.shape[0]

x_noeq_y_right = all_x_noeq_y[all_x_noeq_y == 1].shape[0] / all_x_noeq_y.shape[0]
x_noeq_y_mistake = all_x_noeq_y[all_x_noeq_y == 0].shape[0] / all_x_noeq_y.shape[0]

In [10]:
result = сhance_x_eq_y * x_eq_y_right + сhance_x_noeq_y * x_noeq_y_right

In [11]:
print('Если ответы совпадают:')
print('\t вероятность правильного решения: {:.2%} вероятность ошибки: {:.2%}'.format(
    x_eq_y_right, x_eq_y_mistake))
print('Если ответы не совпадают:')
print('\t вероятность правильного решения: {:.2%} вероятность ошибки: {:.2%}'.format(
    x_noeq_y_right, x_noeq_y_mistake))

Если ответы совпадают:
	 вероятность правильного решения: 95.61% вероятность ошибки: 4.39%
Если ответы не совпадают:
	 вероятность правильного решения: 50.00% вероятность ошибки: 50.00%


In [12]:
print('Шанс менеджера получить верный результат при таком подходе: {:.2%}'.format(result))

Шанс менеджера получить верный результат при таком подходе: 82.36%


##### Рассмотрим  второй вариант при котором стажеры-аналитики совершают разные ошибки при решении задач, тогда  

In [13]:
# вероятность что решения совпадают: оба решили правильно (два решения одинаковые)
сhance_x_eq_y = ver_true_x*ver_true_y
# вероятность что решения не совпадают: первый решил верно второий нет + оба ошиблись (два разных решения)
сhance_x_noeq_y = ver_true_x*ver_false_y + ver_false_x*ver_true_y + ver_false_x*ver_false_y

In [14]:
print('вероятность того что решения совпадают: {:.2%}\nрешения не совпадают: {:.2%}'.format(сhance_x_eq_y, сhance_x_noeq_y))

вероятность того что решения совпадают: 67.83%
решения не совпадают: 32.17%


In [15]:
all_x_eq_y = x[(x==y) & (x==1)]
all_x_noeq_y = x[(x!=y) | ((x==y) & (x==0))]

In [16]:
x_eq_y_right = all_x_eq_y[all_x_eq_y == 1].shape[0] / all_x_eq_y.shape[0]
x_eq_y_mistake = all_x_eq_y[all_x_eq_y == 0].shape[0] / all_x_eq_y.shape[0]

x_noeq_y_right = all_x_noeq_y[all_x_noeq_y == 1].shape[0] / all_x_noeq_y.shape[0]
x_noeq_y_mistake = all_x_noeq_y[all_x_noeq_y == 0].shape[0] / all_x_noeq_y.shape[0]

In [17]:
result = сhance_x_eq_y * x_eq_y_right + сhance_x_noeq_y * x_noeq_y_right

In [18]:
print('Если ответы совпадают:')
print('\t вероятность правильного решения: {:.2%} вероятность ошибки: {:.2%}'.format(
    x_eq_y_right, x_eq_y_mistake))
print('Если ответы не совпадают:')
print('\t вероятность правильного решения: {:.2%} вероятность ошибки: {:.2%}'.format(
    x_noeq_y_right, x_noeq_y_mistake))

Если ответы совпадают:
	 вероятность правильного решения: 100.00% вероятность ошибки: 0.00%
Если ответы не совпадают:
	 вероятность правильного решения: 45.16% вероятность ошибки: 54.84%


In [19]:
print('Шанс менеджера получить верный результат при таком подходе: {:.2%}'.format(result))

Шанс менеджера получить верный результат при таком подходе: 82.36%


#### В двух рассмотренных случаях, при данной стратегии вероятность получения правильного ответа не изменилось и составила 82.36%, что соответствует исходной вероятности каждого из стажеров-аналитиков. Второй стажер-аналитик не влияет на увеличения вероятности получения правильного решения.